### DISCUSSION TITLE
Flushing of metrics in PeriodicReader

### DISCUSSION THREAD

**Discussion by Achooo** [open-telemetry/opentelemetry-go#3993] [Q&A] [answered]
Hi team :)

I've been diving into the code for the PeriodicReader (see [here](https://github.com/open-telemetry/opentelemetry-go/blob/1b55281859bfaf4b03b7968a059c5239cf8c7a20/sdk/metric/periodic_reader.go#L178)). I noticed that the `collectAndExport` function only exports metrics using the client. It does not clear any stored metrics from my understanding. The flush operation seems to only relate to client export.

I'm trying to understand, after reading through the overall SDK code (pipelines, instruments, etc.), how does instrumented metric data _in memory_ get flushed/replaced? 

Thanks a lot!

Edit: Think I figured parts of it out!

- A  Reader - i.e. a PeriodicReader  will collectAndExport metrics every X seconds ([code](https://github.com/open-telemetry/opentelemetry-go/blob/37388599eb3602000025f64ab6c0cca18c848582/sdk/metric/periodic_reader.go#L173))
- When it collects metrics, it calls the produce function of a pipeline ([code](https://github.com/open-telemetry/opentelemetry-go/blob/37388599eb3602000025f64ab6c0cca18c848582/sdk/metric/periodic_reader.go#L262))
- The pipeline calls the Aggregation method on the aggregators ([code](https://github.com/open-telemetry/opentelemetry-go/blob/37388599eb3602000025f64ab6c0cca18c848582/sdk/metric/pipeline.go#L163))
- The different aggregators, when Aggregation is called, return the aggregated data and delete **existing data -> when metrics get flushed from memory essentially**
  - Code example for [LastValue](https://github.com/open-telemetry/opentelemetry-go/blob/37388599eb3602000025f64ab6c0cca18c848582/sdk/metric/internal/lastvalue.go#L71) or [DeltaSum](https://github.com/open-telemetry/opentelemetry-go/blob/37388599eb3602000025f64ab6c0cca18c848582/sdk/metric/internal/sum.go#L94))
  - I’m a bit confused why the CummulativeSum doesn’t do so ([code](https://github.com/open-telemetry/opentelemetry-go/blob/37388599eb3602000025f64ab6c0cca18c848582/sdk/metric/internal/sum.go#L132))
  
Does this understanding seem correct?

**Comment by MrAlias** [comment, marked as answer]
> I’m a bit confused why the CummulativeSum doesn’t do so (code)

Cumulative temporality means data from all collection cycles is exported. It is not flushed from memory due to this essential behavior it needs to provide.

**Reply by Achooo** [reply]
Thanks that makes sense! I was expecting it to flush after X cycles or some sort of memory management there. I see this issue is open https://github.com/open-telemetry/opentelemetry-go/issues/3006 to address this.

**Comment by JobberRT** [comment]
Sorry for asking here, but I can't get this to work. I'm using `PeriodRead` + `DeltaTemporality`. But `Prometheus` still says it's "XXX_counter_total" and the data is cummulative. I don't know which part did I do wrong. I also read the [Official Doc](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#metric-points)

## Example Code
### 1. Create the `Exporter`

```golang
func newGRPCExporter(endpoint string) sdkmetric.Exporter {
	exp, err := otlpmetricgrpc.New(nil,
		otlpmetricgrpc.WithEndpoint(endpoint),
		otlpmetricgrpc.WithCompressor("gzip"),
		otlpmetricgrpc.WithTimeout(15*time.Second),
		otlpmetricgrpc.WithInsecure(),
		otlpmetricgrpc.WithTemporalitySelector(temporalityDeltaSelector),
	)
	if err != nil {
		panic("failed to create grpc exporter")
	}
	return exp
}

func temporalityDeltaSelector(_ sdkmetric.InstrumentKind) metricdata.Temporality {
	return metricdata.DeltaTemporality
}

```

By default, `Counter` will use `SumAggregation`, then I assign `DeltaTemporality` to the exporter
https://github.com/open-telemetry/opentelemetry-go/blob/e2f8ecae74aba268326d2fbf52c6ccce21e56698/sdk/metric/reader.go#L147-L160

### 2. Setup the `MeterProvider` with the `Exporter` I just created

```golang
func setMeterProvider(res *resource.Resource, exp sdkmetric.Exporter, interval time.Duration) {
	var meterProvider *sdkmetric.MeterProvider
	meterProvider = sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(
			exp,
			sdkmetric.WithTimeout(15*time.Second),
			sdkmetric.WithInterval(interval)),
		),
	)
	otel.SetMeterProvider(meterProvider)
	providerShutdown = meterProvider.Shutdown
}
```

### 3. Do some load

```golang
var (
	counter metric.Int64Counter
)

func do() {
	counter.Add(context.Background(), rand.Int63n(10))
}

// metrics.Meter() just a wrapper function of otel.Meter()
func ExecuteMetrics() {
	counter, _ = metrics.Meter().Int64Counter("test counter")
	for i := 0; i < 10; i++ {
		time.Sleep(time.Duration(rand.Intn(6000)) * time.Millisecond)
		fmt.Println("do!")
		do()
	}
}
```

### 4. Finally shutdown the `MeterProvider`

## Prometheus behavior
<img width="1286" alt="image" src="https://github.com/open-telemetry/opentelemetry-go/assets/17498767/f42a7855-1dea-457e-958a-8cd9cb44ae10">

## What I except to happen
Counter will be reset every export. For the demo, the data is randomly generated within 10 so I think in `Prometheus` the line chart should never beyond 10, and the line randomly go up or down.

I know there's a `UpDownCounter` but I think the `UpDownCounter` is mainly used for counter's data could add or sub, not the graph can go up or down(Within a export loop, data in SDK can be add or sub).

**Reply by MrAlias** [reply]
Prometheus only works with cumulative temporality. I'm guessing you are exporting to a collector that is then being scraped by Prometheus. The collector is performing a delta-to-cumulative transform to ensure Prometheus gets the correct data.

**Reply by Achooo** [reply]
Exactly as @MrAlias mentioned, Prometheus doesn't support Delta. Counters should be aggregated continuously (no resets) in a cummulative fashion.

In our collector, the otel contrib's `prometheusremotewrite` that we use will even drop any metrics that aren't cumulative (https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/db8208c23af6f5ac8782b2c2c241ac16fcf36131/pkg/translator/prometheusremotewrite/helper.go#L247) 

We ended up changing the temporality to cumulative when using the Metrics SDK.

Some useful links:
- https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#sums
- https://opentelemetry.io/docs/specs/otel/metrics/data-model/#temporality

**Comment by JobberRT** [comment]
Sorry for replying via email if this is inconvenient to read.

Then could I use Gauge instead？Before using Otel sdk， I was using this to
achieve what I need.

Will try tomorrow, got off work today


Tyler Yahn ***@***.***> 于 2023年6月1日周四 22:19写道：
